Fix/raw

AaronO opened the pull request:
This PR fixes #447 

### Summary

* [x] Add unit tests for broken blocks in raw
* [x] Implement a fix in `src/parser.js`

### Bonus
* [x] Add tests for raw blocks in raw blocks
* [x] Support raw blocks in raw blocks

Ideally I think this should be fixed in `src/lexer.js` at the Tokenizer level. (Maybe I'm wrong on that, someone with a better knowledge of the codebase should know).

SamyPesse commented:
:+1:

adamhavel commented:
Does this fix a similar problem with broken Nunjucks comments in raw block? For example:

```
{% raw %}{# test {% endraw %}
```

This statement throws an error, too.

AaronO commented:
@adamhavel The `Tokenizer` is fairly strict, which generally is a good thing, since it makes errors easier to spot and fix.

The problem is that `raw` blocks use the `Tokenizer` so it actually tries to semantically parse the contents of the raw block which triggers all these errors since it's strict and we really don't care about the syntax validity inside raw blocks, the ideal solution would be to not use the tokenizer inside a `raw` block and be lower level.

However it's not that simple because the rest of the parser operates on top of the Tokenizer, so making `raw` blocks an exception isn't easy given the current structure and abstractions in place.

I'll have to dive into the code and figure out how to refactor it in a way that's clean and improves the robustness of `{% raw %}` blocks, but somebody on the `nunjucks` team definitely has a better knowledge of the codebase, so it would be great if thy could jump in.

AaronO commented:
@adamhavel I added a unit test for your case in my code.

And more importantly I rewrote `Parser.parseRaw` so that it's a lot more robust and in fact far simpler (it's well commented and is still 10 lines shorter than the original while being far more robust).

It passes all the unit tests and we shouldn't have any strange issues with raw blocks again.

AaronO commented:
@jlongster This should be good to review.

jlongster commented:
@AaronO thanks a bunch! currently overloaded because of a work week coming up (next week), but I'll try to get through this soon

AaronO commented:
@jlongster Have you had a chance to take a look at it yet ?

Let me know if you have any questions. Overall this PR simplifies the `raw` block support and makes it more robust, I don't really think there are any regressions (I added extra tests, it passes all those and the old ones of course).

It would be nice to move back on to the `master`: https://github.com/GitbookIO/gitbook/blob/master/package.json#L16 ;)

jlongster commented:
@AaronO Haven't had the best time maintaining open source projects this year, sorry. This looks great though! Merging.

AaronO commented:
@jlongster No problem, it happens to all of us at times :) Thanks for merging the PR ! :+1:
